core/helpers.nix

```
@@ -1,25 +1,57 @@
 # override to the standard nixpkgs lib, adding
 # terranix-related helpers
 pkgs: self: super: {
-  tf = {
+  tf = rec {
     # escape ${} to create a terraform reference string
     ref = ref: "\${${ref}}";
 
     # easier way to create templatefile() invocations
     # if used to set a variable, remember to wrap in `ref`
     template =
-      { text ? ""
-      , source ? ""
-      , variables ? { }
+      {
+        text ? "",
+        source ? "",
+        variables ? { },
       }:
-        assert super.assertMsg (text == "" || source == "") "You must provide either 'text' or 'source' arguments";
-        let
-          template =
-            if text != ""
-            then pkgs.writeText "template.tftpl" text
-            else source;
-        in
-        "templatefile(\"${template}\", ${builtins.toJSON variables})";
+      assert super.assertMsg (
+        text == "" || source == ""
+      ) "You must provide either 'text' or 'source' arguments";
+      let
+        template = if text != "" then pkgs.writeText "template.tftpl" text else source;
+      in
+      "templatefile(\"${template}\", ${builtins.toJSON variables})";
+
+    # Apply function to all strings in an attrset recursively through lists as well
+    # Converts stringlike types to string before (derivations become paths for example)
+    mapStringsDeep =
+      function: value:
+      if super.isStringLike value then
+        function (toString value)
+      else if super.isAttrs value then
+        super.mapAttrs (name: mapStringsDeep function) value
+      else if super.isList value then
+        map (mapStringsDeep function) value
+      else
+        value;
+
+    # Escapes strings so variables aren't interpolated with ${} and %{}
+    escapeEscape =
```

Shouldn't this one be called `escapeInterpolation`?

The ["escape escape" sequences](https://developer.hashicorp.com/terraform/language/expressions/strings#escape-sequences) are `$${}` and `%%{}`

```suggestion
    escapeInterpolation =
```